File: services/rust/salvo/src/routes/middleware/chaos.rs
use app::state::AppState;
use rand::{RngExt, rng};
use salvo::{Depot, FlowCtrl, Request, Response, http::StatusCode, writing::Json};
use serde_json::json;
use std::{
    sync::{Arc, Mutex},
    time::Duration,
};
use tokio::time::sleep;

#[salvo::handler]
pub async fn chaos_middleware(
    req: &mut Request,
    depot: &mut Depot,
    res: &mut Response,
    ctrl: &mut FlowCtrl,
) {
    let Ok(state) = depot.obtain::<Arc<Mutex<AppState>>>().cloned() else {
        tracing::error!(
            method = %req.method(),
            path = %req.uri(),
            "Missing app state in chaos_middleware"
        );
        res.stuff(
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(json!({"error": "Internal server error"})),
        );
        return;
    };

    let (latency_enabled, latency_rate, error_enabled, error_rate) = match state.lock() {
        Ok(guard) => (
            guard.config.latency.enabled,
            guard.config.latency.rate,
            guard.config.error.enabled,
            guard.config.error.rate,
        ),
        Err(_) => {
            tracing::error!(
                method = %req.method(),
                path = %req.uri(),
                "Poisoned lock in chaos_middleware"
            );
            res.stuff(
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({"error": "Internal server error"})),
            );
            return;
        }
    };

    if latency_enabled && rng().random_range(0..=100) < latency_rate {
        let duration = Duration::from_micros(rng().random_range(500..=1500));
        sleep(duration).await;
    }

    if error_enabled && rng().random_range(0..=100) < error_rate {
        res.stuff(
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(json!({"error": "Internal server error"})),
        );
        return;
    }

    ctrl.call_next(req, depot, res).await;
}

File: services/go/chi/routes/middleware/chaos.go
package middleware

import (
	"app"
	"chi/routes"
	"math/rand/v2"
	"net/http"
	"time"
)

func ChaosMiddleware(state *app.AppState) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			latencyEnabled := state.Config.Latency.Enabled
			latencyRate := state.Config.Latency.Rate
			errorEnabled := state.Config.Error.Enabled
			errorRate := state.Config.Error.Rate

			if latencyEnabled && rand.IntN(101) < int(latencyRate) {
				delay := time.Duration(500+rand.IntN(1001)) * time.Microsecond
				time.Sleep(delay)
			}
			if errorEnabled && rand.IntN(101) < int(errorRate) {
				routes.AppErrors.Internal.Error(w)
				return
			}

			next.ServeHTTP(w, r)
		})
	}
}
